### ISSUE TITLE
Node metrics for autoscaler

### ISSUE THREAD

**Issue by killianmuldoon** [DirectXMan12/k8s-prometheus-adapter#180]
I'm using prometheus adapter alongside the node exporter, but I can't seem to scrape the node metrics from the custom metrics API. My scrape config is: 

```yaml
    - seriesQuery: '{__name__=~"^node_.*",job="kubernetes-service-endpoints"}'
      seriesFilters: []
      resources: {overrides: {system_name: {resource: node}}}
      name:
        matches: ^node_(.*) 
      metricsQuery: avg_over_time(<<.Series>>[30s])
```
When I query 
`kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1`
 I get a lot of node metrics returned, but when I query the specific metric using
`kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1│
/nodes/*/memory_SwapTotal | jq .`
I get back 

```json
{                                                                           
"kind": "MetricValueList", 
"apiVersion": "custom.metrics.k8s.io/v1beta1",                            
"metadata": {       
    "selfLink": "/apis/custom.metrics.k8s.io/v1beta1/nodes/%2A/memory_SwapTotal
},          
  "items": []                                                      
}
```

This happens for every single node metric as far as I can tell. The metrics appear properly in Prometheus.

In the Prometheus Adapter log files when I make this query I can see the call being made to prometheus and returning in data. So most of the pipeline seems to be functioning. On my one node cluster I get 

```
{
"status":"success",
"data":{
  "resultType":"vector",
  "result":[{"metric": 
    {"addonmanager_kubernetes_io_mode":"Reconcile",
     "instance":"10.237.208.164:9100",
     "job":"kubernetes-service-endpoints",
     "kubernetes_io_cluster_service":"true",
     "kubernetes_io_name":"NodeExporter",
     "kubernetes_name":"node-exporter",
     "kubernetes_namespace":"kube-system"},
     "value":[1553190789.726,"624859136"]}]}
}
```
so I think there's some issue that I can't figure out with my Prometheus Adapter config. 

I tried to follow the solution in #123, but it doesn't seem to have worked.

**Comment by xjas** [comment]
Maybe the resource overrides you used can not meet your meet? 

You have tried this `kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1`, can it return the metric `nodes/memory_SwapTotal` ?

I wonder, can you try this config?https://github.com/DirectXMan12/k8s-prometheus-adapter/issues/123#issuecomment-438052778
```
- seriesQuery: '{__name__=~"^node_.*",node!=""}'  # match things with a `node` label that start with `node_`
  resources: {overrides: {node: node}}  # the only resource we expose this on is nodes, using the `node` label
  name: {matches: "^node_(.*)$"}  # metrics like node_memory_active become nodes/memory_active in the custom metrics API
  metricsQuery: "sum(<<.Series>>{<<.LabelMatchers>>}) by (<<.GroupBy>>)"  # since this isn't a rate, just aggregate by whatever resource we request it on (i.e. nodes right now)
```

**Comment by xjas** [comment]
I think the request you used is wrong, it's not `kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1/nodes/*/memory_SwapTotal | jq .` , it should be `kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1/node/*/memory_SwapTotal | jq .` . 

The difference is `nodes/*/memory_SwapTotal` and `node/*/memory_SwapTotal`.

If not lucky, the above is still not work, could you provide the results of ` kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/" | jq .`

**Comment by killianmuldoon** [comment]
So either way, with nodes or node, it returns the same output. The configuration used from #123 doesn't return any node metrics whatsoever unfortunately. 

The output of "/apis/custom.metrics.k8s.io/v1beta1/" | jq  lists all of the metrics that I'm looking for, but querying those metrics results in the output described in the OP. I don't have access to the machine right now so I can't post the full output unfortunately.

**Comment by xjas** [comment]
If neither nodes nor node can return a expected output, I think there maybe a bug as I described in issue #178.
I guess the below is part of the output when you try `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/" | jq . `.

```
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "custom.metrics.k8s.io/v1beta1",
  "resources": [
    {
      "name": "nodes/memory_SwapTotal",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    }
  ]
}
```

Looking carefully into the result, If you find the field `namespaced` is `true`, you probably can't get the metric you want. In fact, the resource nodes is root-scoped apparently. And I think this is a bug.

Another way to prove it: namespace is also a root-scoped resource, you can try `kubectl get --raw /apis/custom.metrics.k8s.io/v1beta1│ /namespace/*/memory_SwapTotal | jq .` if you have exposed the resource in your config rules. This may return some meaningful results.

If that happened as I described, this could be a bug which I have mentioned in #178, the relevant PR is #179.

There are a few amends that I think probably is right. If you want to try, you can pull the branch in my repository.

**Comment by xjas** [comment]
I have build an image for your convenience, and you can try `docker pull xjas/k8s-prometheus-adapter-amd64`

**Comment by killianmuldoon** [comment]
Thanks for the help xjas! I got access to the machine, but it's not giving results inline with your approach.

[custom_metrics_raw_list.txt](https://github.com/DirectXMan12/k8s-prometheus-adapter/files/3000429/custom_metrics_raw_list.txt)

So all of my node metrics have namespaced: false as you can see in the result of the get attached. I don't think your fix will work for that, unfortunately. 

Querying them under namespace/*/ gives 
`Error from server (NotFound): the server could not find the metric memory_SwapTotal for namespace`

**Comment by killianmuldoon** [comment]
Hey, I managed to get this working at last. The issue wasn't with namespacing, it was with incorrect labelling in my prometheus configuration i.e. my node metrics were being picked up from a generic "kubernetes service endpoint" and weren't being labelled properly as a result. 

Thanks again for the help xjas!

**Comment by xjas** [comment]
If possible, could you let me know what the output is when you use the official Adapter image: directxman12/k8s-prometheus-adapter-amd64:v0.4.0 or version v0.4.1?

In fact, I wonder, which image you are using about the custom_metrics_raw_list.txt you attached in https://github.com/DirectXMan12/k8s-prometheus-adapter/issues/180#issuecomment-475966601?

Because Looking carefully into the `custom_metrics_raw_list.txt` you attached, I see `"namespaced": false` about resource nodes, take `nodes/node_memory_VmallocUsed` for an example.
```
    {
      "name": "nodes/node_memory_VmallocUsed",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
```
The output you pasted did not consist with what I get when querying `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/" | jq .`. In fact when I use image `directxman12/k8s-prometheus-adapter-amd64:v0.4.0 or version v0.4.1`, I always get the field `"namespaced": true,` of resource nodes. As a contrast, I can get the field `"namespaced": false,` of resource namespaces. 

I think this inconsistency may be a bug, as I descirbed in issue #178 and PR #179.
```
{
  "kind": "APIResourceList",
  "apiVersion": "v1",
  "groupVersion": "custom.metrics.k8s.io/v1beta1",
  "resources": [
    {
      "name": "nodes/reflector_short_watches",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "namespaces/storage_operation_errors",
      "singularName": "",
      "namespaced": false,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    },
    {
      "name": "nodes/container_cpu_cfs_throttled_periods",
      "singularName": "",
      "namespaced": true,
      "kind": "MetricValueList",
      "verbs": [
        "get"
      ]
    }
  ]
}
```
If I was right, then this probably is a bug, otherwise I just didn't use it correctly. So could I have your confirmation?

If you have tried the image `xjas/k8s-prometheus-adapter-amd64:latest`, could you help pasting the output of  `kubectl get --raw "/apis/custom.metrics.k8s.io/v1beta1/" | jq .` when using the image `xjas/k8s-prometheus-adapter-amd64:latest`as well?



Sorry for my verbose descirbing. If possible, I want to know the following things to confirm the probable bug.
1. which image you are using about the custom_metrics_raw_list.txt you attached in https://github.com/DirectXMan12/k8s-prometheus-adapter/issues/180#issuecomment-475966601?
2. if you have tried, what's the output when using image `xjas/k8s-prometheus-adapter-amd64:latest`?

Thank you in advance.

**Comment by killianmuldoon** [comment]
So the image in my case is
 `directxman12/k8s-prometheus-adapter-amd64` 

I haven't gotten a chance to use the image in your pull request just yet. Where are the metrics you're focused on coming from ? Are they per-node versions of pod metrics, or are they from the node exporter or similar? That might be the differentiating factor.

**Comment by xjas** [comment]
Thanks for your quick response.
I just try the `directxman12/k8s-prometheus-adapter-amd64:latest` images, but the return is still `"namespaced": true`.
Sorry, I don't understand the meaning of what you say; 
> Are they per-node versions of pod metrics, or are they from the node exporter or similar? That might be the differentiating factor.

The metric I need is `http_requests_total
Could you make a few explain or give me some references, thanks.

**Comment by killianmuldoon** [comment]
So the metrics I've been looking at come from the Prometheus Node Exporter https://github.com/prometheus/node_exporter
I'm capturing them with a specific scrape job on prometheus. I don't see http_requests_total there, so I think you're getting your metrics from a difference source. Do you know what Prometheus scrape job produces the metric?

**Comment by xjas** [comment]
Thanks for your response,
The metrics I used did not come form Prometheus Node Exporter, I think the metric comes from a  Prometheus scrape job too. In fact it's exported by the cadvisor.

I will do some deep dig, and I will let you know if I get some infomation.
Thanks again.

**Comment by killianmuldoon** [comment]
I think I have one metric exposed by the cadvisor job and your solution of using namespace rather than node did in fact return the metrics. The metrics I'm interested in though are from the node exporter. The way that you scrape and relabel through the prometheus config map might be the key there.

**Comment by xjas** [comment]
Thanks for your advice.
Forgive my stupidity, what you have said is about adapter rules config and the Prometheus config? Did I understand you correctly?

Thanks for your patience, and I will try harder.

**Comment by killianmuldoon** [comment]
No need for forgiveness! I should explain myself better. So the solution to my problem, which might be applicable in your case, was in the prometheus scrape config. I'm not sure if it will help you or not. 
```
      - job_name: 'kubernetes-node-exporter'
        tls_config:
          ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
          insecure_skip_verify: true
        bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
        scheme: https
        kubernetes_sd_configs:
        - role: node
        relabel_configs:
        - source_labels: [__address__]
          regex: ^(.*):\d+$
          target_label: __address__
          replacement: $1:9100
        - target_label: __scheme__
          replacement: http
       # Here's where I add the host name label
        - source_labels: [__meta_kubernetes_node_name]
          target_label: none
```
In the relabel config above I added the kubernetes node name to the list. this is what made the metrics appear properly to prometheus adapter in my case. It may be possible to do something similar in your case, but I'm not certain. If your solution is working for you know it's probably a good idea to stick with it!

**Comment by xjas** [comment]
I appreciate your consideration.

I am trying it as you have advised and I think it must help a lot to make my adapter works correctly.

If I have some progress, I will let you know.

Thanks for your detailed explanation again.

**Comment by xjas** [comment]
The metrics I used is indeed `per-node versions of pod metrics`, which is different from what you are using.
So I need PromL do the aggregation for me, and the resource nodes have to be root-scoped.

Recent progress is PR #179 has been merged for now, however, the shortcoming of this PR is that it's something like hard-code. If any root-scoped resources is introduced into K8S, the code need to be modified again.

Fortunately, there is a plan to `unifies detection of namespaced objects` as described in https://github.com/DirectXMan12/k8s-prometheus-adapter/pull/179#issuecomment-477567312.

Thanks for your hint sincerely again.
